exchanges/bybit/bybit.go

```diff
@@ -0,0 +1,746 @@
+	}{}
+	err := by.SendAuthHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, bybitPastOrder, params, &resp, privateSpotRate)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Data, nil
+}
+
+// GetWalletBalance returns user wallet balance
+func (by *Bybit) GetWalletBalance(ctx context.Context) ([]Balance, error) {
+	resp := struct {
+		Data struct {
+			Balances []Balance `json:"balances"`
+		} `json:"result"`
+	}{}
+	err := by.SendAuthHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, bybitWalletBalance, url.Values{}, &resp, privateSpotRate)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Data.Balances, nil
+}
+
+// SendHTTPRequest sends an unauthenticated request
+func (by *Bybit) SendHTTPRequest(ctx context.Context, ePath exchange.URL, path string, f request.EndpointLimit, result interface{}) error {
+	endpointPath, err := by.API.Endpoints.GetURL(ePath)
+	if err != nil {
+		return err
+	}
+	item := &request.Item{
+		Method:        http.MethodGet,
+		Path:          endpointPath + path,
+		Result:        result,
+		Verbose:       by.Verbose,
+		HTTPDebugging: by.HTTPDebugging,
+		HTTPRecording: by.HTTPRecording}
+
+	return by.SendPayload(ctx, f, func() (*request.Item, error) {
+		return item, nil
+	})
+}
+
+// SendAuthHTTPRequest sends an authenticated HTTP request
+func (by *Bybit) SendAuthHTTPRequest(ctx context.Context, ePath exchange.URL, method, path string, params url.Values, result interface{}, f request.EndpointLimit) error {
+	endpointPath, err := by.API.Endpoints.GetURL(ePath)
+	if err != nil {
+		return err
+	}
+
+	recvWindow := int64(5000) // default millisecond
+	if params.Get("recvWindow") != "" {
+		var recvWindowParam int64
+		// convert recvWindow value into time.Duration
+		recvWindowParam, err = convert.Int64FromString(params.Get("recvWindow"))
+		if err != nil {
+			return err
+		}
+		recvWindow = time.Duration(recvWindowParam).Milliseconds()
+	}
+	params.Set("recvWindow", strconv.FormatInt(recvWindow, 10))
+	params.Set("timestamp", strconv.FormatInt(time.Now().UnixMilli(), 10))
+	params.Set("api_key", by.API.Credentials.Key)
+	signature := params.Encode()
+	hmacSigned, err := crypto.GetHMAC(crypto.HashSHA256, []byte(signature), []byte(by.API.Credentials.Secret))
+	if err != nil {
+		return err
+	}
+	hmacSignedStr := crypto.HexEncodeToString(hmacSigned)
+
+	headers := make(map[string]string)
+	var payload []byte
+	switch method {
+	case http.MethodPost:
+		headers["Content-Type"] = "application/json"
+		m := make(map[string]string)
+		m["api_key"] = by.API.Credentials.Key
+
+		for k, v := range params {
+			m[k] = strings.Join(v, "")
+		}
+		m["sign"] = hmacSignedStr
+		payload, err = json.Marshal(m)
+		if err != nil {
+			return err
+		}
+	default:
+		headers["Content-Type"] = "application/x-www-form-urlencoded"
+		path = common.EncodeURLValues(path, params)
+		path += "&sign=" + hmacSignedStr
+		endpointPath += path
+	}
+
+	return by.SendPayload(ctx, f, func() (*request.Item, error) {
```

Comment:
Apart from endpoint path we will need to generate a new signature on every retry, so the majority of this code will need to be added to the function signature so that will correctly be handled by the requester.